Sources/Basic/TemporaryFile.swift

```diff
@@ -0,0 +1,117 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import libc
+import func POSIX.getenv
+import class Foundation.NSFileHandle
+import var Foundation.NSUTF8StringEncoding
+
+public enum TempFileError: ErrorProtocol {
+    /// Could not create a unique temporary filename.
+    case couldNotCreateUniqueName
+    /// Some error thrown defined by posix's open().
+    // FIXME: This should be factored out into a open error enum.
+    case other(Int32)
+}
+
+private extension TempFileError {
+    init(errno: Int32) {
+        switch errno {
+        case libc.EEXIST:
+            self = .couldNotCreateUniqueName
+        default:
+            self = .other(errno)
+        }
+    }
+}
+
+/// This class is basically a wrapper over posix's mkstemps() function to creates disposable files.
+/// The file is deleted as soon as the object of this class is deallocated.
+public final class TemporaryFile {
+    /// If specified during init, the temporary file name begins with this prefix.
+    public let prefix: String
+
+    /// If specified during init, the temporary file name ends with this suffix.
+    public let suffix: String
+
+    /// The directory in which the temporary file is created.
+    public let dir: String
+
+    /// The full path of the temporary file.
+    public let path: String
+
+    /// The file descriptor of the temporary file. It is used to create NSFileHandle which is exposed to clients.
+    private let fd: Int32
+
+    /// FileHandle of the temporary file, can be used to read/write data.
+    public let fileHandle: NSFileHandle
+
+    /// Creates an instance of Temporary file. The temporary file will live on disk until the instance
+    /// goes out of scope.
+    ///
+    /// - Parameters:
+    ///     - dir: If specified the temporary file will be created in this directory otherwise enviornment variables
+    ///            TMPDIR, TEMP and TMP will be checked for a value (in that order). If none of the env variables are
+    ///            set, dir will be set to `/tmp/`.
+    ///     - prefix: The prefix to the temporary file name.
+    ///     - suffix: The suffix to the temporary file name.
+    ///
+    /// - Throws: TempFileError
+    public init(dir: String? = nil, prefix: String = "", suffix: String = "") throws {
+        self.suffix = suffix
+        self.prefix = prefix
+        // Determine in which directory to create the temporary file.
+        self.dir = TemporaryFile.determineTempDirectory(dir)
+        // Construct path to the temporary file.
+        let path = self.dir + prefix + "TemporaryFile.XXXXXX" + suffix
```

Comment:
I'd prefer making `TemporaryFile` the default prefix, than embed it in every temporary path.